src/basicfuns.jl

```diff
@@ -152,9 +152,62 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(_x::Real)
+    x = float(_x)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x < x0
+        return exp(x)
+    elseif x < x1
+        return log1p(exp(x))
+    elseif x < x2
+        return x + exp(-x)
+    else
+        return x
+    end
+end
+
+#= The precision of BigFloat cannot be computed from the type only.
+Since computing the dynamically thresholds is slow, prefer to use
+a slower but accurate implementation in this case. =#
+function log1pexp(x::BigFloat)
+    if x > 0
+        return x + log1p(exp(-x))
+    else
+        return log1p(exp(x))
+    end
+end
+
+#=
+Returns thresholds x0, x1, x2 such that:
+
+    * log1pexp(x) ≈ exp(x) for x ≤ x0
+    * log1pexp(x) ≈ log1p(exp(x)) for x0 < x ≤ x1
+    * log1pexp(x) ≈ x + exp(-x) for x1 < x ≤ x2
+    * log1pexp(x) ≈ x for x > x2
+
+where the tolerances of the approximations are on the order of eps(typeof(x)).
+For types for which `precision(x)` depends only on the type of `x`, the compiler
+should optimize away all computations done here.
+=#
+@inline function _log1pexp_thresholds(x::Real)
+    z = (precision(x) - 1) * oftype(float(x), IrrationalConstants.logtwo)
+    x0 = -z - IrrationalConstants.logtwo
+    x1 = z / 2
```

Comment:
We don't need `float(x)`, the function is only called with floating point number types. One could also avoid the subtraction:
```suggestion
    prec = precision(x)
    logtwo = oftype(x, IrrationalConstants.logtwo)
    x0 = - prec * logtwo
    x1 = ((prec - 1) * logtwo) / 2
```

Reply:
Agree with the removing `float(x)`. For the computations, it seems you are eliminating a subtraction but introducing more calculations in `x0, x1`.

Reply:
Yeah. Ideally, there would be a constant for `log(sqrt(2))` that we could use, then it would lead to the much simpler code
```suggestion
    prec = precision(x)
    x0 = - prec * oftype(x, IrrationalConstants, logtwo)
    x1 = (prec - 1) * oftype(x, IrrationalConstants.logsqrt2)
```

Reply:
Ended up liking this version since the subtraction is between ints which is exact. Marking resolved.